File: classes.py
import matplotlib.pyplot as plt
import numpy as np
from random import choice

class chemical:
    default_id = 64

    def __init__(self,diff_const=1,id="",color=""):
        if id == "":
            chemical.default_id = chemical.default_id + 1
            self.id = chr(chemical.default_id)
        else:
            self.id = id
        self.diff_const = diff_const

        if color == "":
            self.color = '#' + ''.join([choice('0123456789ABCDEF') for j in range(6)])
        else:
            self.color = color

    def __repr__(self):
        return self.id

class reaction:
    def __init__(self,chems_in:dict,chems_out:dict,reaction_const=1):
        self.chems_in = chems_in.copy()
        self.chems_out = chems_out.copy()
        self.chems_list = set(list(chems_in) + list(chems_out))
        self.reaction_const = reaction_const

        self.id = ""
        first = True
        if self.chems_in:
            for chem in self.chems_in:
                if first:
                    self.id = self.id + str(self.chems_in[chem]) + chem.id + " "
                    first = False
                else:
                    self.id = self.id + "+ " + str(self.chems_in[chem]) + chem.id + " "
        else:
            self.id = self.id + "Ø "

        self.id = self.id + "-(" + str(reaction_const) + ")> "

        first = True
        if self.chems_out:
            for chem in self.chems_out:
                if first:
                    self.id = self.id + str(self.chems_out[chem]) + chem.id + " "
                    first = False
                else:
                    self.id = self.id + "+ " + str(self.chems_out[chem]) + chem.id + " "
        else:
            self.id = self.id + "Ø"

    def __repr__(self):
        return self.id

    def has_chem(self,chem:chemical):
        return chem in self.chems_list

    def pred_prey(self,chems_val:dict):
        c = self.reaction_const
        for chem in self.chems_in:
            c = c * chems_val[chem]**self.chems_in[chem]
    
        # consumption
        consumption = {}
        # if not self.chems_out:
        #     for chem in self.chems_in:
        #         consumption[chem] = self.reaction_const*self.chems_in[chem]
        for chem in self.chems_list:
            if chem in self.chems_in:
                if chem in self.chems_out:
                    if self.chems_in[chem] > self.chems_out[chem]:
                        val = self.chems_in[chem] - self.chems_out[chem]
                    else:
                        val = 0
                else:
                    val = self.chems_in[chem]
            else:
                val = 0
            consumption[chem] = c * val
            
        # creation
        creation = {}
        if not self.chems_in:
            for chem in self.chems_out:
                creation[chem] = self.reaction_const*self.chems_out[chem]
        for chem in self.chems_list:
            if chem in self.chems_out:
                if chem in self.chems_in:
                    if self.chems_in[chem] < self.chems_out[chem]:
                        val = self.chems_out[chem] - self.chems_in[chem]
                    else:
                        val = 0
                else:
                    val = self.chems_out[chem]
            else:
                val = 0
            creation[chem] = c * val

        return consumption, creation

class rod:
    rod_num = 0

    def __init__(self,dt,dx,length=1,temp=1):
        self.chems_list = set()
        self.chems = {}
        self.chems_next = {}
        self.reacts = []
        self.dx = dx
        self.dt = dt
        self.nx = int(length/dx)
        self.length = length
        self.temp = temp
        self.coef = dt/(dx**2)
        self.x = []
        for i in range(self.nx):
            self.x.append(i*self.dx)

        self.fig, self.ax = plt.subplots()
        rod.rod_num += 1
        self.fig.canvas.manager.set_window_title("Rod " + str(rod.rod_num))
        plt.ion()

    def __repr__(self):
        out = "=================================\nChemicals\n=================================\n"
        # chemicals
        for chem in self.chems_list:
            out = out + chem.id + " "
            for i in self.chems[chem]:
                out = out + str(i) +" "
            out = out + "\n"

        out = out + "=================================\nReactions\n=================================\n"
        #reactions
        for react in self.reacts:
            out = out + react.id + "\n"

        return out

    def add_reaction(self,reacts):
        if type(reacts) == reaction:
            self.reacts.append(reacts)
        elif type(reacts) == list:
            self.reacts = self.reacts + reacts

    def add_chem(self,chem:chemical,region:list,vals:list):
        if len(region) != len(vals):
            raise Exception("Provide region and vals of the same size")
        
        if not chem in self.chems_list:
            self.chems_list.add(chem)
            self.chems[chem] = [0] * self.nx
            self.chems_next[chem] = [0] * self.nx

        for i in range(len(region)):
            self.chems[chem][region[i]] = self.chems[chem][region[i]] + vals[i]
        
    def transfer_to(self,space,chem,range_in=[],range_out=[]):
        # to do: add surface transfer support
        
        if range_in == []:
            range_in = list(range(self.nx))

        if range_out == []:
            range_out = range_in.copy()

        transfer_val = 0
        for i in range(len(range_in)):
            transfer_val = transfer_val + self.chems[chem][i]
            self.chems[chem][range_in[i]] = 0

        transfer_val = [transfer_val/len(range_out)] * len(range_out)
        space.add_chem(chem,range_out,transfer_val)

    def integrate_chems(self,chems=None) -> dict:
        integrated = {}
        if chems == None or type(chems) == list:
            if chems == None:
                chems = self.chems_list

            for chem in chems:
                integrated[chem] = 0
                for i in self.chems[chem]:
                    integrated[chem] = integrated[chem] + i*self.dx

        elif type(chems) == chemical:
            chem = chems
            integrated[chem] = 0
            for i in self.chems[chem]:
                integrated[chem] = integrated[chem] + i*self.dx

        else:
            raise Exception("Give no argumant for all chemicals, list of chemicals wanted, or single chemical")
        
        return integrated

    def update(self):
        for i in range(1,self.nx-1):
            cons = {}
            cre = {}
            for react in self.reacts:
                if set(react.chems_list).issubset(set(self.chems_list)):
                    chems_val = {}
                    for chem in react.chems_in:
                        chems_val[chem] = self.chems[chem][i]
                    cons_temp, cre_temp = react.pred_prey(chems_val)
                    for j in cons_temp:
                        if j in cons:
                            cons[j] = cons[j] + cons_temp[j]
                            cre[j] = cre[j] + cre_temp[j]
                        else:
                            cons[j] = cons_temp[j]
                            cre[j] = cre_temp[j]
                else:
                    raise Exception("Missing chemicals in rod for reaction")
            for chem in self.chems_list:
                if not chem in cons:
                    cons[chem] = 0
                if not chem in cre:
                    cre[chem] = 0
                
                self.chems_next[chem][i] = self.coef*chem.diff_const*(self.chems[chem][i-1] - 2*self.chems[chem][i] + self.chems[chem][i+1])\
                    + self.chems[chem][i]\
                    + self.dt*(cre[chem]- cons[chem])

        for chem in self.chems_list:
            self.chems_next[chem][0] = self.chems_next[chem][1]
            self.chems_next[chem][self.nx-1] = self.chems_next[chem][self.nx-2]

        for chem in self.chems_list:
            self.chems[chem] = self.chems_next[chem].copy()

    def show(self,time=None,ylim=True,pause=0.00000001):
        self.ax.clear()
        if time != None:
            self.fig.suptitle("Time: " + str(time))
        else:
            self.fig.suptitle("")
        self.ax.set_xlabel("x")
        self.ax.set_ylabel("Chemical Concentration")
        for chem in self.chems_list:
            self.ax.plot(self.x,self.chems[chem],color=chem.color,label=repr(chem))
        self.fig.legend()
        if type(ylim) != bool:
            self.ax.set_ylim(ylim[0],ylim[1])
        self.fig.canvas.draw()
        self.fig.show()
        plt.pause(pause)

class surface:
    surface_num = 0

    def __init__(self,dx,dy,dt,length_x=1,length_y=1,temp=1):
        self.chems_list = set()
        self.chems = {}
        self.chems_next = {}
        self.reacts = []
        self.dx = dx
        self.dy = dy
        self.dt = dt
        self.nx = int(length_x/dx)
        self.ny = int(length_y/dy)
        self.length_x = length_x
        self.length_y = length_y
        self.temp = temp 
        self.coef = None #####################################################
        self.x = np.arange(0,1,self.dx)
        self.y = np.arange(0,1,self.dy)
        self.x, self.y = np.meshgrid(self.x,self.y)

        self.fig, self.ax = plt.subplots()
        surface.surface_num += 1
        self.fig.canvas.manager.set_window_title("Surface " + str(surface.rod_num))
        plt.ion()

    def __repr__(self):
        pass

    def add_reaction(self,react:reaction):
        self.reacts.append(react)

    def add_chem(self,chem:chemical,region:list,vals:list):
        if len(region) != len(vals):
            raise Exception("Provide region and vals of the same size")
        
        if not chem in self.chems_list:
            self.chems_list.add(chem)
            self.chems[chem] = [0] * self.nx
            self.chems_next[chem] = [0] * self.nx

        for i in range(len(region)):
            self.chems[chem][region[i]] = self.chems[chem][region[i]] + vals[i]

    def transfer_to(self,space,chem,range_in=[],range_out=[]):
        pass

    def integrate_chems(self,chems=None) -> dict:
        pass

    def update(self):
        pass

    def show(self,time=None):
        pass

File: src/classes.cpp
#include <classes.h>
#include <helpers.h>
#include <iostream>
#include <string>
#include <math.h>
#include <stdexcept>
#include <algorithm>

Chemical::Chemical() {
    default_id++;
    id = char(default_id);
    diff_const = 1.0;
}

Chemical::Chemical(double diff_const_param, std::string id_param) {
    id = id_param;
    diff_const = diff_const_param;
}

size_t Chemical::ChemicalHashFunction::operator()(const Chemical& chemical) const {
    return std::hash<std::string>{}(chemical.id);
}

bool Chemical::ChemicalEqualFunction::operator()(const Chemical& lhs, const Chemical& rhs) const {
    return lhs.id == rhs.id && lhs.diff_const == rhs.diff_const;
}

bool Chemical::operator==(const Chemical& other) const {
    return id == other.id;
}

bool Chemical::operator!=(const Chemical& other) const {
    return !(*this == other);
}

int Chemical::default_id = 64;

Reaction::Reaction(ChemicalMap chems_in_param, ChemicalMap chems_out_param, double reaction_const_param) {
    chems_in = chems_in_param;
    chems_out = chems_out_param;

    for (auto chem : chems_in) {
        chems_list.push_back(chem.first);
    }
    for (auto chem : chems_out) {
        for (auto chem_in : chems_list) {
            if (!(chem.first.id == chem_in.id && chem.first.diff_const == chem_in.diff_const)) {
                chems_list.push_back(chem.first);
            }
        }
    }

    reaction_const = reaction_const_param;

    id = "";
    bool first = true;
    if(!chems_in.empty()) {
        for (auto chem : chems_in) {
            Chemical c = chem.first;
            if (first) {
                id = id + double_to_string(chems_in[c], 2) + c.id + " ";
                first = false;
            } else id = id + "+ " + double_to_string(chems_in[c], 2) + c.id + " ";
        }
    } else id = id + "Ø ";

    id = id + "-(" + double_to_string(reaction_const, 2) + ")> ";

    first = true;
    if (!chems_out.empty()) {
        for (auto chem : chems_out) {
            Chemical c = chem.first;
            if (first) {
                id = id + double_to_string(chems_out[c], 2) + c.id + "";
                first = false;
            } else id = id + "+ " + double_to_string(chems_out[c], 2) + c.id + " ";
        }
    } else id = id + "Ø ";
}

std::pair<ChemicalMap, ChemicalMap> Reaction::pred_prey(ChemicalMap chems_val) {
    double c = reaction_const;
    for (std::pair<const Chemical, double> chem_in : chems_in) {
        Chemical chem = chem_in.first;
        c = c * pow(chems_val[chem], chems_in[chem]);
    }

    //consumption
    ChemicalMap consumption;
    double val;
    for (Chemical chem : chems_list) {
        if (chems_in.find(chem) != chems_in.end()) {
            if (chems_out.find(chem) != chems_out.end()) {
                if (chems_in[chem] > chems_out[chem]) val = chems_in[chem] - chems_out[chem];
                else val = 0.0;
            } else val = chems_in[chem];
        } else val = 0.0;
        consumption[chem] = c * val;
    }

    //creation
    ChemicalMap creation;
    val = 0;
    if (chems_in.empty()) {
        for (std::pair<const Chemical, double> chem_out : chems_out) {
            Chemical chem = chem_out.first;
            creation[chem] = reaction_const * chems_out[chem];
        }
    }
    for (Chemical chem : chems_list) {
        if (chems_out.find(chem) != chems_out.end()) {
            if (chems_in.find(chem) != chems_in.end()) {
                if (chems_in[chem] < chems_out[chem]) val = chems_out[chem] - chems_in[chem];
                else val = 0.0;
            } else val = chems_out[chem];
        } else val = 0.0;
        creation[chem] = c * val;
    }

    std::pair<ChemicalMap, ChemicalMap> ret;
    ret.first = consumption; 
    ret.second = creation;
    return ret;
}

Rod::Rod(double dt_param, double dx_param, double length_param, double temp_param) {
    dx = dx_param;
    dt = dt_param;
    length = length_param;
    nx = int(length/dx);
    temp = temp_param;
    coef = dt/(dx*dx);
    for (int i = 0; i < nx; i++) {
        x.push_back(i*dx);
    }
}

void Rod::add_reaction(Reaction react) {
    reactions.push_back(react);
}

void Rod::add_reaction(std::vector<Reaction> reacts) {
    reactions.insert(reactions.end(), reacts.begin(), reacts.end());
}

void Rod::add_chem(Chemical chem, std::vector<int> region, std::vector<double> vals) {
    if (region.size() != vals.size()) throw std::invalid_argument("Provide region and vals of the same size");

    if (std::find(chems_list.begin(), chems_list.end(), chem) == chems_list.end()) {
        chems_list.push_back(chem);
        std::vector<double> v(nx, 0.0), vn(nx, 0.0);
        chems[chem] = v;
        chems[chem] = vn;
    }

    for (int i = 0; i < region.size(); i++) {
        chems[chem][region[i]] += vals[i];
    }
}

void Rod::transfer_to(Rod destination, Chemical chem, std::vector<int> range_in, std::vector<int> range_out) {
    if (range_in.empty()) {
        for (int i = 0; i < nx; i++) range_in.push_back(i); 
    }

    if (range_out.empty()) {
        for (int i = 0; i < destination.nx; i++) range_out.push_back(i); 
    }

    double transfer_vol = 0;
    for (int i = 0; i < range_in.size(); i++) {
        transfer_vol += chems[chem][i];
        chems[chem][range_in[i]] = 0;
    }

    // Different from transfer_vol
    std::vector<double> transfer_vals;
    for (int i = 0; i < range_out.size(); i++) transfer_vals[i] = transfer_vol/range_out.size();
    destination.add_chem(chem, range_out, transfer_vals);
}

ChemicalMap Rod::integrate_chems(Chemical chem) {
    ChemicalMap integrated;
    integrated[chem] = 0;
    for (double value : chems[chem]) integrated[chem] += value*dx;

    return integrated;
}

ChemicalMap Rod::integrate_chems(std::vector<Chemical> chems_int) {
    ChemicalMap integrated;
    if (chems_int.empty()) chems_int = chems_list;

    for(Chemical chem : chems_int) {
        integrated[chem] = 0;
        for (double value : chems[chem]) integrated[chem] += value*dx;
    }

    return integrated;
}

void Rod::update() {
    for (int i = 1; i < nx-1; i++) {
        ChemicalMap cons, cre;
        for (Reaction react : reactions) {
            if (is_subset(react.chems_list, chems_list)) {
                ChemicalMap chems_val;
                std::vector<Chemical> c_i;
                for (std::pair<Chemical, double> p : react.chems_in) c_i.push_back(p.first);
                for (Chemical chem : c_i) chems_val[chem] = chems[chem][i];
                std::pair<ChemicalMap, ChemicalMap> temp = react.pred_prey(chems_val);
                ChemicalMap cons_temp = temp.first, cre_temp = temp.second;
                for (std::pair<Chemical, double> j : cons_temp) {
                    cons[j.first] += cons_temp[j.first];
                    cre[j.first] += cre_temp[j.first];
                }
            } else throw std::runtime_error("Missing chemicals in rod for reaction");
        }
        for (Chemical chem : chems_list) {
            if (cons.find(chem) == cons.end()) cons[chem] = 0.0;
            if (cre.find(chem) == cre.end()) cre[chem] = 0.0;

            chems_next[chem][i] = coef*chem.diff_const*(chems[chem][i-1] - 2*chems[chem][i] + chems[chem][i+1])
                                  + chems[chem][i]
                                  + dt*(cre[chem] - cons[chem]);
        }
    }

    // Boundary conditions
    for (Chemical chem : chems_list) {
        chems_next[chem][0] = chems_next[chem][1];
        chems_next[chem][nx-1] = chems_next[chem][nx-2];
    }

    // Move chems_next into chems
    for (Chemical chem : chems_list) for (int i = 0; i < chems.size(); i++) chems[chem][i] = chems_next[chem][i];
}